### PR TITLE
RFC: Improve std.random.unpredictableSeed

### PR DESCRIPTION
**Pull request by yshui**
Implemented according to suggestions from Cédric Picard (see [1](https://cym13.github.io/article/unpredictableSeed.html) [2](https://forum.dlang.org/post/atpkaykjsfiolghtivfq@forum.dlang.org))

I only have linux machine to test with, so I opened this pull request to abuse the build bot.

Shoot if you spot any problems.

### CODE REVIEW

**Comment by wilzbach** [comment]
> so I opened this pull request to abuse the build bot.

That's what it's for.

For Linux >= 3.17` there's the awesome [`getrandom`](http://man7.org/linux/man-pages/man2/getrandom.2.html) system call which should be used. For more details please see this [intro article](https://lwn.net/Articles/605828)

Also you might be very interested in [my related work at mir.random](https://github.com/libmir/mir-random/pull/13) which also exposes a `genRandom{,Blocking}` API.

File: std/random.d [hunk 1]
```diff
@@ -1281,17 +1294,76 @@ A single unsigned integer seed value, different on each successive call
...
+    {
+        import core.sys.windows.wincrypt;
+        static HCRYPTPROV hProv;
+        static bool initialized;
+        uint rand;
+        if (!initialized)
+        {
+            bool ret = CryptAcquireContext(&hProv, null, null, PROV_RSA_FULL, 0);
+            if (!ret)
+                throw new Exception("CryptAcquireContext");
```
Comment (wilzbach) [inline std/random.d]: it's possible to recover from this:

```
            if (GetLastError() == NTE_BAD_KEYSET)
            {
                 // Attempt to create default container
                 if (!CryptAcquireContextA(&hProvider, null, null, PROV_RSA_FULL, CRYPT_NEWKEYSET | CRYPT_SILENT))
                    return 1;
```

File: std/random.d [hunk 2]
```diff
@@ -1281,17 +1294,76 @@ A single unsigned integer seed value, different on each successive call
...
-    if (!seeded)
+    version(Windows)
+    {
+        import core.sys.windows.wincrypt;
+        static HCRYPTPROV hProv;
+        static bool initialized;
+        uint rand;
+        if (!initialized)
+        {
+            bool ret = CryptAcquireContext(&hProv, null, null, PROV_RSA_FULL, 0);
```
Comment (wilzbach) [inline std/random.d]:
```d
CryptAcquireContextW(&hProv, null, null, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
```

-> https://msdn.microsoft.com/en-us/library/windows/desktop/aa379942(v=vs.85).aspx

**Comment by wilzbach** [comment]
It took me _a very long time_ (and many runs on AppVeyor) to figure out that the Wincrypt headers in druntime aren't correct for x64. Hence I highly recommend to have use the overrides I use in Mir - it's tested ;-)

```d
// the wincrypt headers in druntime are broken for x64!
private alias ULONG_PTR = size_t; // uint in druntime
private alias BOOL = bool;
private alias DWORD = size_t; // uint in druntime
private alias LPCWSTR = wchar*;
private alias PBYTE = ubyte*;
private alias HCRYPTPROV = ULONG_PTR;
private alias LPCSTR = const(char)*;

private extern(Windows) BOOL CryptGenRandom(HCRYPTPROV, DWORD, PBYTE) @nogc @safe nothrow;
private extern(Windows) BOOL CryptAcquireContextA(HCRYPTPROV*, LPCSTR, LPCSTR, DWORD, DWORD) @nogc nothrow;
private extern(Windows) BOOL CryptAcquireContextW(HCRYPTPROV*, LPCWSTR, LPCWSTR, DWORD, DWORD) @nogc nothrow;
private extern(Windows) BOOL CryptReleaseContext(HCRYPTPROV, ULONG_PTR) @nogc nothrow;

private __gshared ULONG_PTR hProvider;
```

**Comment by yshui** [comment]
@wilzbach What's the correct way to detect Linux version in D?

**Comment by yshui** [comment]
@wilzbach Maybe you should push changes to druntime?

**Comment by wilzbach** [comment]
> @wilzbach What's the correct way to detect Linux version in D?

There's no "correct" way, but the system call [`uname`](http://man7.org/linux/man-pages/man2/uname.2.html) is a good option.

-> Another friendly redirect to my work: https://github.com/libmir/mir-random/pull/13/files
Copied here for convenience:

```d
import core.sys.posix.sys.utsname : utsname;

// druntime isn't properly annotated
private extern(C) int uname(utsname* __name) @nogc nothrow;

private bool hasGetRandom() @nogc @trusted nothrow
{
    import core.stdc.string : strtok;
    import core.stdc.stdlib : atoi;

    utsname uts;
    uname(&uts);
    char* p = uts.release.ptr;

    // poor man's version check
    auto token = strtok(p, ".");
    int major = atoi(token);
    if (major  > 3)
        return true;

    if (major == 3)
    {
        token = strtok(p, ".");
        if (atoi(token) >= 17)
            return true;
    }

    return false;
}
```

> @wilzbach Maybe you should push changes to druntime?

I haven't used Windows for the last five years, so I don't know much about the Windows bindings in druntime.

File: std/random.d [hunk 3]
```diff
@@ -1270,6 +1270,19 @@ alias Xorshift    = Xorshift128;                            /// ditto
...
 }
 
+version(OpenBSD)
+{
+    version = ARC4_RANDOM;
+}
+else version(NetBSD)
+{
+    version = ARC4_RANDOM;
+}
```
Comment (jacob-carlborg) [inline std/random.d]: macOS also has the `arc4random` function. Not sure what's best to use.

[1] https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man3/arc4random.3.html

File: std/random.d [hunk 4]
```diff
@@ -1281,17 +1294,79 @@ A single unsigned integer seed value, different on each successive call
 */
 @property uint unpredictableSeed() @trusted
 {
-    import core.thread : Thread, getpid, MonoTime;
-    static bool seeded;
-    static MinstdRand0 rand;
-    if (!seeded)
+    version(Windows)
```
Comment (JackStouffer) [inline std/random.d]: space between version and `(`

File: std/random.d [hunk 5]
```diff
@@ -1281,17 +1294,79 @@ A single unsigned integer seed value, different on each successive call
 */
 @property uint unpredictableSeed() @trusted
```
Comment (JackStouffer) [inline std/random.d]: this should be a template with trusted escapes so the attributes can be inferred

File: std/random.d [hunk 6]
```diff
@@ -1270,6 +1270,19 @@ alias Xorshift    = Xorshift128;                            /// ditto
...
+    version = ARC4_RANDOM;
+}
+else version(NetBSD)
+{
+    version = ARC4_RANDOM;
+}
+
+version(ARC4_RANDOM)
+{
+    extern(C) uint arc4random();
```
Comment (JackStouffer) [inline std/random.d]: `@nogc nothrow pure`

File: std/random.d (hunk 3 above)
Comment (yshui) [inline std/random.d]: In recent version of OpenBSD/NetBSD, arc4random is supported by ChaCha20. On macOS it's ARC4.

OTH macOS's /dev/random uses Yarrow, so I guess that's a better choice?

File: std/random.d (hunk 6 above)
Comment (yshui) [inline std/random.d]: I don't think this can be pure?

File: std/random.d [hunk 7]
```diff
@@ -1270,6 +1270,19 @@ alias Xorshift    = Xorshift128;                            /// ditto
...
 
+version (OpenBSD)
+{
+    version = ARC4_RANDOM;
+}
+else version (NetBSD)
+{
+    version = ARC4_RANDOM;
+}
+
```
Comment (n8sh) [inline std/random.d]: I suggest also setting `version = ARC4_RANDOM` for `version (FreeBSD)` and `version (OSX)`.

**Comment by n8sh** [comment]
On my machine I profiled all of the implementations except the one using CryptGenRandom:

-  The current `unpredictableSeed` takes 2.9 seconds for 100 million calls.
- `arc4random()` took 3.2 seconds for 100 million calls. I ran this test under OS X whose arc4random uses AES rather than ChaCha20.
- The implementation that reads from `/dev/urandom` on every call takes around a minute and ten seconds for 100 million calls.
- The fallback implementation based on `SHA256Digest` takes nine minutes for 100 million calls.

I think the last two have the wrong speed/unpredictability tradeoff for a non-cryptographic API.

**Comment by JackStouffer** [comment]
Going to close this in a week if @yshui doesn't respond.

**Comment by yshui** [comment]
@JackStouffer I'm not sure where this pull request is going...

**Comment by JackStouffer** [comment]
@yshui

* Does it matter how slow `unpredictableSeed` is (I suspect no, because you should really only be calling it once per run of the software)?
* Does `arc4random` work on OSX and FreeBSD?
* Add info to the docs about the crypto security of the seed on each platform
* Add a warning that this function shouldn't be re-called for every new random number created and should really just be called once per run

Also, @wilzbach you seem to know about this area, any other comments on the implementation?

**Comment by n8sh** [comment]
>Does `arc4random` work on OSX and FreeBSD?

OSX has arc4random with AES instead of ChaCha20.

FreeBSD has an arc4random that [still uses](https://github.com/freebsd/freebsd/blob/master/lib/libc/gen/arc4random.c) RC4 and has the problem that forked processes see the same sequences when the PID wraps. It shouldn't be used anywhere a cryptographically secure PRNG is required, but for non-cryptographic uses it would still be a significant improvement over the current `unpredictableSeed`.

**Comment by JackStouffer** [comment]
Ping @yshui

**Comment by quickfur** [comment]
One general comment / question:  reading the articles posted at the two given links in the original PR description, I come away with the understanding that what the article was advocating for was *not* to try to make `unpredictableSeed` cryptographically secure, but rather to point out that `std.random` was never intended to be used for cryptographic purposes in the first place, and therefore user code should *not* be relying on `std.random` for any cryptographic properties it may or may not have.

In that context, this PR seems a little misplaced.  What's the value of going through all of this trouble to use `/dev/urandom` and/or its equivalent to get a supposedly cryptographically-secure seed value, only to use it to seed a PRNG that was never designed for cryptographic applications in the first place, and that is therefore almost certainly insecure?

**Comment by yshui** [comment]
@quickfur I agree. I have the impression that this issue has already been solved when the disclaimer is add to the std.random document.

I will close this PR.

**Pull request closed without merging**